partner_vat_unique/models/res_partner.py

```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Grant Thornton Spain - Ismael Calvo <ismael.calvo@es.gt.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, models, _
+from odoo.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    @api.constrains('vat')
+    def _check_vat_unique(self):
+        results = self.env['res.partner'].read_group([
+            ('parent_id', '=', False),
+            ('vat', 'in', map(lambda p: p.vat, filter(lambda p: p.vat, self)))
+        ], ['vat'], ['vat'])
```

Comment:
Yikes, this is so hard to read that you possibly didn't notice its bugs. What is exactly the criteria you want to follow?

Bulk-checking vats would probably lead you to weird corner cases where editing 1 record in a batch that raises a false positive... Better just iterate over records and check one by one.

I guess you want to only check partners that have no parent, so first of all document that behavior, which might not be so obvious. Then start by excluding those that have a parent:

```python
for record in self:
    if record.parent_id:
        # Contacts use parent's VAT, so they are not checked
        continue
```


Next, you want to check a record doesn't exist, but you forgot to add current record's id exclusion:

```python
domain = [('parent_id', '=', False), ('vat', '=', record.vat), ('id', '!=', record.id)]
```